pkg/agent/multicast/mcast_discovery.go

```diff
@@ -134,13 +141,28 @@ func (s *IGMPSnooper) validate(event *mcastGroupEvent) (bool, error) {
 	return true, nil
 }
 
-func (s *IGMPSnooper) validatePacketAndNotify(event *mcastGroupEvent) {
-	allow, _ := s.validate(event)
-	if allow {
-		s.eventCh <- event
+func (s *IGMPSnooper) addToIGMPReportNPStatsMap(item types.McastNPValidationItem, packetLen uint64) {
+	uuid := item.Uuid
+	name := item.Name
+	ruleType := *item.NPType
+	if s.igmpReportNPStats[ruleType][uuid] != nil {
```

Comment:
You did not answer my previous question:

> This is accessed by a single goroutine, and we need not locking?

Reply:
Updated. Locking is needed.

Reply:
After having a second thought, I think locking is not needed. There is only one goroutine modifying content of `igmpReportNPStats`(spawned by "MulticastGroupDiscovery" packetIn handler here).

https://github.com/antrea-io/antrea/blob/05504f0337df267f092cfccee2cd1c89a603557e/pkg/agent/openflow/packetin.go#L110

It should be not a problem for the case that antrea-agent is writing a packet metric to `igmpReportNPStats` while users are reading it. Users may get stats with or without this packet metric for now, but that's fine for stats applications.

The only concern here is the inconsistency between packet count and byte count, which I resolved by making "writing one RuleMetric to igmpReportNPStats" atomic.

Not sure I have answered your question. Thanks

Reply:
First, are we sure packetIn handler is a single routine? @wenyingd 

Second, after you return igmpReportNPStats to readers, they will iterate/look up the map and its nested maps? That is not safe when the maps can be modified by another routine.

Reply:
> First, are we sure packetIn handler is a single routine? @wenyingd
Yes, the packetIn handler is a single routine. Please refer to https://github.com/antrea-io/antrea/blob/main/pkg/agent/openflow/packetin.go#L105  For IGMP report message, we actually share reason value 1 with ANP. So a single goroutine processes the packet-in messages for both ANP and Multicast.

As for the reason to share value 1, it is because we found an issue that other reasons set in the flow actiosn except for 0 or 1 are not identified in the packet-in messages while receiving from OVS.

Reply:
Updated the implementation, igmpReportNPStats is deep copied first then get returned.